[HEAVY DNM] Adding Energy Weapons

lquinn2015 opened the pull request:
### Energy Weapons TODO issue #613 

### Approach
Added 8 prefabs. FusionCells, LaserPistol, LaserGun, Tazer, 2 Sound manager changes/prefabs, LaserShot, TazerShot. 

Edited Weapon.cs  -> added a toggle for spawning bullet casings. Note i set the default for it to true as to not step on ballistic weapons.

Edited WeaponNetworkActions.cs  -> to use the toggled listed above.

Created BulletEnergyShot.cs -> while this class is empty now i am sure it will expand due to different damage types, maybe it will denote bombs etcs. 

Add the sounds to the sound manager in the death match scene and applied them. 

### Open Questions and Pre-Merge TODOs

- [x]  I read the contribution guidelines and the wiki.
- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  This PR does not include files without specific need to do so.
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer

Additonals
- [x]  Removing map changes
- [x]  Sounds Changed
- [x]  Tested.
- [x] Todo in bullet Behavior script left


## The TODO note
So i left a todo in bullet behavior cs because the damage and damage type will likely need to be set on the call of OnShoot.

### Notes:
I have a bad eye for the guns inhands numbers they should look close enough however. Also my tazer sound would likely work for touching raw wires.

lquinn2015 committed:
EnergyWeapons

lquinn2015 committed:
EnergyWeapons

lquinn2015 committed:
AddedEnergyWeaponsToNULLSEC

lquinn2015 committed:
Merge remote-tracking branch 'upstream/develop' into develop

lquinn2015 committed:
Revert "Merge remote-tracking branch 'upstream/develop' into develop"

This reverts commit 7bb953e419a487db788ccc944f36f7996e40528e, reversing
changes made to b71860f8db28b9ec7078826dbb8568646c82e702.

lquinn2015 committed:
Revert "Merge remote-tracking branch 'upstream/develop' into develop"

This reverts commit 7bb953e419a487db788ccc944f36f7996e40528e, reversing
changes made to b71860f8db28b9ec7078826dbb8568646c82e702.

Crow-Control reviewed (changes requested):
Could you remove the map changes, as those need to be done by an actual mapper to preserve the mapping balance and design in the whole map.

Further more we are going to get it into 0.2.2 with a bit of luck, but that also mean that you are on our most stringent bitch and test list for this pr. Sorry :P

lquinn2015 committed:
FinalUpdate

Fixes Tazer Sound
Has been Tested ie no related bugs.
Sends Burn damage via BULLET (temp saved for a rewrite of some bullet behavior)

Crow-Control reviewed (changes requested):
You have included some files related to your test build (the DLL) please remove those and report back before I start a full re-review

lquinn2015 committed:
Dll Removed

Crow-Control commented on UnityProject/Assets/Scripts/Health/Living/PlayerHealth.cs:
```diff
@@ -220,6 +220,10 @@ protected override void OnDeathActions()
+				// FIXME
+				// reenable killfeed once the systemmessages are fixed
+				
+				/*
```
This code was removed for a reason, why does your PR re add this old code?!

Crow-Control commented on UnityProject/Assets/Scripts/Health/Living/PlayerHealth.cs:
```diff
@@ -266,7 +270,7 @@ protected override void OnDeathActions()
 					//Combat demo killfeed
 					//PostToChatMessage.Send(killerName + " has killed " + gameObject.name, ChatChannel.System);
-				}
+				}*/
```
Same here

Crow-Control commented on UnityProject/Assets/Scripts/Managers/GameData.cs:
```diff
@@ -140,7 +140,7 @@ private void OnLevelFinishedLoading(Scene scene, LoadSceneMode mode)
 	private IEnumerator WaitToStartServer()
 	{
-		yield return new WaitForSeconds(0.1f);
+		yield return new WaitForSeconds(5f);
```
Again, this was changed some time ago and you just revert it.

Crow-Control commented on UnityProject/Assets/Scripts/Managers/GameManager.cs:
```diff
@@ -4,20 +4,19 @@
 using UnityEngine;
```
You just reverted all changed that where done in this file the past few days

Crow-Control commented on UnityProject/Assets/Scripts/Messages/Client/ClientMessage.cs:
```diff
@@ -1,20 +1,15 @@
+		SentBy = LocalPlayerId();
 		CustomNetworkManager.Instance.client.connection.Send(GetMessageType(), this);
-		Debug.LogFormat("Sent {0}", this);
+		//		Debug.LogFormat("Sent {0}", this);
```
well I forgot to turn this debug off thanks for noticing, but still the rest shouldn't be reverted by your PR

Crow-Control commented on UnityProject/Assets/Scripts/Messages/Client/PostToChatMessage.cs:
```diff
@@ -15,19 +15,12 @@ public class PostToChatMessage : ClientMessage
 	public override IEnumerator Process()
 	{
 		yield return WaitFor(SentBy);
-		if (NetworkObject)
```
Again, reverting changes others made you shouldn't even be touching

Crow-Control commented on UnityProject/Assets/Scripts/Messages/Server/UpdateRoundTimeMessage.cs:
```diff
@@ -14,7 +14,7 @@ public override IEnumerator Process()
 		yield return WaitFor(Subject);
 
-		GameManager.Instance.SyncTime(Time);
+		GameManager.Instance.SyncTimendResetCounter(Time);
```
Again, don't do this

Crow-Control commented on UnityProject/Assets/Scripts/PlayGroups/PlayerScript.cs:
```diff
@@ -244,20 +244,17 @@ public ChatChannel GetAvailableChannels(bool transmitOnly = true)
 			//TODO add missing modifiers
-			//TODO add if for being drunk
-			//ChatModifier modifiers = ChatModifier.Drunk;
+			ChatModifier modifiers = ChatModifier.Drunk;
```
and so on...

Crow-Control commented on UnityProject/Assets/Scripts/Weapons/Projectiles/BulletBehaviour.cs:
```diff
@@ -27,7 +28,17 @@ private void StartShoot(Vector2 dir, float angle, GameObject controlledByPlayer,
+     * This function should be converted to a standard non abstract function as to  
+     *      Move forward with making the guns be the deteriminers of damage. This means you will
+     *      go by Gun's damage value and type rather then the bullet. Note we probably want to store/use the weapon that
+     *      does the firing here some how. 
```
The gun already determines the damage, As I told you.
The gun gives the damage value to the bullet. 

So this should be removed as it's wrong. Bullets don't have a a damage value of themselves, look at the gun prefab for damage values

Crow-Control commented on UnityProject/Assets/Scripts/Weapons/Projectiles/BulletEnergyShot.cs:
```diff
@@ -0,0 +1,8 @@
+    public override void OnShoot()
+    {
+        damageType = DamageType.BURN; // sets damage hook to burn damage these are energy weapons
+        //Bullet specific stuff
```
Why the "bullet specific stuff" explain what need to be added and use //FIXME or //TODO as a prefix ;)

Crow-Control commented on UnityProject/Assets/Scripts/Weapons/Weapon.cs:
```diff
@@ -51,10 +51,15 @@ public class Weapon : PickUpTrigger
+        /// <summary>
+        ///     The amount of times per second this weapon can fire
+        /// </summary>
+        public double FireRate;
```
We enforce the use of Tabs instead of spaces, you continuesly use spaces, which we don't want.

Crow-Control commented on UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs:
```diff
@@ -94,9 +94,12 @@ public void CmdUnloadWeapon(GameObject weapon)
-		//TODO add a check to see if bullet or energy weapon
-		ItemFactory.SpawnItem(casingPrefab, transform.position, transform.parent);
-		if (!isFlashing)
+        if (wepBehavior.SpawnsCaseing)
```
Where did you add this "spawnsCaseing flag"? I cant find it anywhere in your code. Or is it already there?

Crow-Control reviewed (changes requested):
Lots of work to be done.
Key points:
- [ ] Prevent reverting other people's commits
- [ ] remove map changes
- [ ] Replace tabs with spaces
- [ ] Look at you casing system
- [ ] Check your code commentary
- [ ] read into how damage gets added to bullets
- [ ] test it.

lquinn2015 committed:
Revert "Revert "Merge remote-tracking branch 'upstream/develop' into develop""

This reverts commit ef20d84115e851923b213207806b21d436576253.

lquinn2015 committed:
Revert "Revert "Merge remote-tracking branch 'upstream/develop' into develop""

This reverts commit ef20d84115e851923b213207806b21d436576253.

lquinn2015 committed:
Revert "Merge remote-tracking branch 'upstream/develop' into develop"

This reverts commit 7bb953e419a487db788ccc944f36f7996e40528e, reversing
changes made to b71860f8db28b9ec7078826dbb8568646c82e702.

lquinn2015 committed:
Revert "Merge remote-tracking branch 'upstream/develop' into develop"

This reverts commit 7bb953e419a487db788ccc944f36f7996e40528e, reversing
changes made to b71860f8db28b9ec7078826dbb8568646c82e702.

lquinn2015 committed:
Merge branch 'develop' of https://github.com/unitystation/unitystation into develop

lquinn2015 committed:
Merge branch 'develop' of https://github.com/unitystation/unitystation into develop

lquinn2015 committed:
Fix merges maybe must test

lquinn2015 committed:
psuh

lquinn2015 committed:
Revert "Revert "Merge remote-tracking branch 'upstream/develop' into develop""

This reverts commit f00bb6fe86e535b8eadde1d2e5b43b3528fde4d9.
